### ISSUE TITLE
Wrapping reentrant call in a loop causes deadlock

### ISSUE THREAD

**Issue by alekw911** [neosmart/AsyncLock#18]
I have taken the sample from the "AsyncLock usage example" section of the documentation and added a for loop around the inner reentrant call and observed a deadlock occur

This is the sample code I used

```csharp
using NeoSmart.AsyncLock;
using System.Threading.Tasks;
using System;
using System.Threading;

public class AsyncLockTest
{
    static async Task Main(string[] args)
    {
        new AsyncLockTest().Test();
        Thread.Sleep(Timeout.Infinite);
    }
    AsyncLock _lock = new AsyncLock();

    void Test()
    {
        // The code below will be run immediately (likely in a new thread)
        Task.Run(async () =>
        {
            // A first call to LockAsync() will obtain the lock without blocking
            using (await _lock.LockAsync())
            {
                // A second call to LockAsync() will be recognized as being
                // reentrant and permitted to go through without blocking.

                for(int i=0;i<100;++i)
                {
                    Console.WriteLine($"Starting iteration {i} from {Thread.CurrentThread.ManagedThreadId}");
                    using (await _lock.LockAsync())
                    {
                        Console.WriteLine($"Obtained lock for iteration {i} from {Thread.CurrentThread.ManagedThreadId}");
                        // We now exclusively hold the lock for 1 minute
                        await Task.Delay(TimeSpan.FromMilliseconds(1));
                        Console.WriteLine($"Finished work for iteration {i} from {Thread.CurrentThread.ManagedThreadId}");
                    }
                    Console.WriteLine($"Released lock for iteration {i} from {Thread.CurrentThread.ManagedThreadId}");
                }
            }
        }).Wait(TimeSpan.FromSeconds(30));

        // This call to obtain the lock is made synchronously from the main thread.
        // It will, however, block until the asynchronous code which obtained the lock
        // above finishes.
        using (_lock.Lock())
        {
            // Now we have obtained exclusive access.
            // <Safely perform non-thread-safe operation safely here>
        }
    }
}
```

**Comment by alekw911** [comment]
Yes it was first discovered when looking into a bug in our application
Then a simpler console app was made to try to determine if it could be reproduced in a context unrelated to the application

**Comment by PolarGoose** [comment]
I have looked into this issue.
The simplified reproduction scenario from @alekw911:
```
using NeoSmart.AsyncLock;
namespace LockTest;
internal class Program
{
    static async Task Main(string[] args)
    {
        var _lock = new AsyncLock();
        Log($"Obtain the lock before entering the for loop");
        using (await _lock.LockAsync())
        {
            for (int i = 0; i < 100; ++i)
            {
                Log(i, "Try to obtain the lock");
                using (await _lock.LockAsync()) // <--------- The deadlock happens here at some iteration
                {
                    Log(i, "Obtained the lock. Start waiting for 1ms");
                    await Task.Delay(TimeSpan.FromMilliseconds(1));
                    Log(i, "Waiting finished.");
                }
                Log(i, "Lock released");
            }
            Log("The loop finished successfully. Release the lock");
        }
        Log("The lock is completely released. Exit the main method");
    }

    static void Log(string msg)
    {
        Console.WriteLine($"[ThreadId={Thread.CurrentThread.ManagedThreadId,-2}] {msg}");
    }

    static void Log(int iteration, string msg)
    {
        Console.WriteLine($"[i={iteration,-2},ThreadId={Thread.CurrentThread.ManagedThreadId,-2}] {msg}");
    }
}
```

It indeed deadlocks in 100% of cases:
```
[ThreadId=1 ] Obtain the lock before entering the for loop
[i=0 ,ThreadId=1 ] Try to obtain the lock
[i=0 ,ThreadId=1 ] Obtained the lock. Start waiting for 1ms
[i=0 ,ThreadId=6 ] Waiting finished.
[i=0 ,ThreadId=6 ] Lock released
[i=1 ,ThreadId=6 ] Try to obtain the lock
[i=1 ,ThreadId=6 ] Obtained the lock. Start waiting for 1ms
[i=1 ,ThreadId=10] Waiting finished.
[i=1 ,ThreadId=10] Lock released
[i=2 ,ThreadId=10] Try to obtain the lock
```

The code locks the mutex in one thread and unlocks it in another.
This happens because using `await` doesn't guarantee that the subsequent code will execute on the same thread.
Therefore, in such case `NeoSmart.AsyncLock` starts working as a non-reentrant lock - basically a functional equivalent of [SemaphoreSlim](https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-7.0).

This behavior explains why [SemaphoreSlim](https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-7.0) doesn't support reentrancy. It's also the reason C# doesn't allow you to use the async keyword inside the lock statement, as shown below:
```
lock (obj)
{
    await Task.Delay(TimeSpan.FromMilliseconds(1));
}
```

There is a very nice [discussion on StackOverflow](https://stackoverflow.com/questions/7612602/why-cant-i-use-the-await-operator-within-the-body-of-a-lock-statement) that provides a lot of details why such deadlocks happen.

What's more concerning is that even this basic code is expected to fail:
```
using (await _lock.LockAsync())
{
    using (await _lock.LockAsync()) // <--- Deadlock is supposed to happen here,
                                    // because it is not guranteed that this line
                                    // will be executed in the same thread that
                                    // holds the lock.
    {

    }
}
```

### Conclusion
* It is impossible to properly implement Asynchronous Reentrant lock in C#. There is a very good article about it [Reentrant Async Lock is Impossible in C#](https://itnext.io/reentrant-recursive-async-lock-is-impossible-in-c-e9593f4aa38a) and a [discussion on Reddit](https://www.reddit.com/r/dotnet/comments/rklmby/reentrant_recursive_async_lock_is_impossible_in_c/)
* `NeoSmart.AsyncLock` primitive shouldn't promise reentrancy when a user uses it with the `async` keyword.

**Comment by mqudsi** [comment]
I'm not relying on the task being dispatched on the same thread to enable reentrance, though. I'm using the extremely poorly documented `AsyncLocal` (combined with traditional same-thread detection, but that's mostly for when using AsyncLock in non-async contexts), which *should* be able to detect this, but it's *extremely* tricky. 

The biggest issue is that `AsyncLocal<T>` doesn't really kick in when an async operation is *dispatched*, only when it is awaited.

https://github.com/neosmart/AsyncLock/blob/0649d7145a99545ac8534bb96c990a20351d8d94/AsyncLock/AsyncLock.cs#L26-L32

**Comment by PolarGoose** [comment]
@mqudsi,
> I'm not relying on the task being dispatched on the same thread to enable reentrance

Sorry, my bad.
However, as far as I understood, the article [Reentrant Async Lock is Impossible in C#](https://itnext.io/reentrant-recursive-async-lock-is-impossible-in-c-e9593f4aa38a) paragraph **ExecutionContext** describes why using `AsyncLocal` primitive doesn't solve the problem of reentrancy.

**Comment by mitja-p** [comment]
@alekw911 how did you fix this for your app?

**Comment by PolarGoose** [comment]
> @alekw911 how did you fix this for your app?

Hello @mitja-p,
I'm not the author of this issue, but it seems that there is no fix, as I described in my [previous message](https://github.com/neosmart/AsyncLock/issues/18#issuecomment-1697860395)

What is your use case? Why do you need an async reentrant lock?

**Comment by IvanGit** [comment]
Hello! What about this async reentrant lock implementation? [.NET Fiddle link](http://dotnetfiddle.net/lLLMoV)

**Comment by PolarGoose** [comment]
@IvanGit,

Ok, so you are asking about the usage of [ReentrantAsyncLock](https://github.com/IvanGit/NuExt.System/blob/main/src/System/Threading/ReentrantAsyncLock.cs).
I have rewritten my reproduction scenario using it. This issue isn't reproduced using this class.
```
internal class Program
{
    static async Task Main(string[] args)
    {
        var asyncLock = new ReentrantAsyncLock();
        Log($"Obtain the lock before entering the for loop");
        await asyncLock.AcquireAsync(async () =>
        {
            Log($"Enter the for loop");
            for (int i = 0; i < 100; ++i)
            {
                await asyncLock.AcquireAsync(async () =>
                {
                    Log(i, "Obtained the lock. Start waiting for 1ms");
                    await Task.Delay(TimeSpan.FromMilliseconds(1));
                    Log(i, "Waiting finished.");
                });
            }
        });
    }

    static void Log(string msg)
    {
        Console.WriteLine($"[ThreadId={Thread.CurrentThread.ManagedThreadId,-2}] {msg}");
    }

    static void Log(int iteration, string msg)
    {
        Console.WriteLine($"[i={iteration,-2},ThreadId={Thread.CurrentThread.ManagedThreadId,-2}] {msg}");
    }
}
```

**Comment by PolarGoose** [comment]
@IvanGit,

Oh, by the way, I noticed you are a creator of [NuExt.System](https://github.com/IvanGit/NuExt.System/tree/main)
Respect for making such a complex library work.

**Comment by alekw911** [comment]
Oh cool a different implementation, I had ended up using: https://github.com/microsoft/vs-threading/blob/main/src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

**Comment by matthew-a-thomas** [comment]
I believe this is a more minimal reproduction of the reported issue:

```csharp
var asyncLock = new AsyncLock();
async Task InnerFunctionAsync()
{
    using (await asyncLock.LockAsync())
    {
        //
    }
}
using (await asyncLock.LockAsync())
{
    await InnerFunctionAsync();
    await InnerFunctionAsync();
}
```

[This dotnetfiddle does not print "done"](https://dotnetfiddle.net/1vanXV).

I suspect the root cause is a race condition. Because a small delay [causes the issue to go away](https://dotnetfiddle.net/RqK0EG):

```csharp
var asyncLock = new AsyncLock();
async Task InnerFunctionAsync()
{
    using (await asyncLock.LockAsync())
    {
        //
    }
}
using (await asyncLock.LockAsync())
{
    await InnerFunctionAsync();
    await Task.Delay(TimeSpan.FromMilliseconds(1));
    await InnerFunctionAsync();
}
```

**Comment by mqudsi** [comment]
Thanks for the simplified test case; this makes the problem considerably more tractable! I'll take a look as soon as I get a chance.

**Comment by matthew-a-thomas** [comment]
> I'm not relying on the task being dispatched on the same thread to enable reentrance, though. I'm using the extremely poorly documented `AsyncLocal` (combined with traditional same-thread detection, but that's mostly for when using AsyncLock in non-async contexts), which _should_ be able to detect this, but it's _extremely_ tricky.
> 
> The biggest issue is that `AsyncLocal<T>` doesn't really kick in when an async operation is _dispatched_, only when it is awaited.
> 
> [AsyncLock/AsyncLock/AsyncLock.cs](https://github.com/neosmart/AsyncLock/blob/0649d7145a99545ac8534bb96c990a20351d8d94/AsyncLock/AsyncLock.cs#L26-L32)
> 
> Lines 26 to 32 in [0649d71](/neosmart/AsyncLock/commit/0649d7145a99545ac8534bb96c990a20351d8d94)
> 
>  // An AsyncLocal<T> is not really the task-based equivalent to a ThreadLocal<T>, in that 
>  // it does not track the async flow (as the documentation describes) but rather it is 
>  // associated with a stack snapshot. Mutation of the AsyncLocal in an await call does 
>  // not change the value observed by the parent when the call returns, so if you want to 
>  // use it as a persistent async flow identifier, the value needs to be set at the outer- 
>  // most level and never touched internally. 
>  private static readonly AsyncLocal<long> _asyncId = new AsyncLocal<long>();

I wonder if you'd have an easier time with `AsyncLocal<T>` if your `T` was a reference type instead of a value type? Here's [an example](https://github.com/matthew-a-thomas/cs-reentrant-async-lock/blob/50d2632ac2903293bfadf4c4bf7d364ee960544c/ReentrantAsyncLock/ReentrantAsyncLock.cs#L57) of what I mean. In that example you can see that the `AsyncLocal`'s value comes out through [this property](https://github.com/matthew-a-thomas/cs-reentrant-async-lock/blob/50d2632ac2903293bfadf4c4bf7d364ee960544c/ReentrantAsyncLock/ReentrantAsyncLock.cs#L88) which is part of [the check for reentrance](https://github.com/matthew-a-thomas/cs-reentrant-async-lock/blob/50d2632ac2903293bfadf4c4bf7d364ee960544c/ReentrantAsyncLock/ReentrantAsyncLock.cs#L164). Of course the MIT license in that project lets you borrow ideas like this, so I'm sure the author of that library (whoever that is 😉) wouldn't mind.

I hope this helps!

**Comment by matthew-a-thomas** [comment]
> It is impossible to properly implement Asynchronous Reentrant lock in C#.

@PolarGoose I disagree for two reasons. [Reason one](https://dotnetfiddle.net/GDkn7D). [Reason two](https://dotnetfiddle.net/6WdVy1). 🙂

Since it _is_ possible, and now that the author of `AsyncLock` has a very minimal repro, I'm confident that he will have this issue ironed out in no time.

**Comment by PolarGoose** [comment]
@matthew-a-thomas,

Great work investigating the issue.

>  I disagree for two reasons

Your examples are similar to the working example that I provided here: https://github.com/neosmart/AsyncLock/issues/18#issuecomment-2335183044 .
What do you want to prove with your examples?

**Comment by PolarGoose** [comment]
@matthew-a-thomas,
This is just an example from the abovementioned article that I tried with the `ReentrantAsyncLock` library:
https://dotnetfiddle.net/QrZrPk
If you look into the log, you can see that the Main and the Child tasks got access to the critical section at the same time:
![Image](https://github.com/user-attachments/assets/ded2d4e7-479d-4f57-bab9-fc2d619d5f1f)

**Comment by matthew-a-thomas** [comment]
@PolarGoose 

> I didn't say that it was impossible to make this specific use case work because there are other implementations that would work for this particular use case.

Ah I see. My apologies for misunderstanding you.

> I think the article that I mentioned describes the problem very well. Do you disagree with it?

Yes. I'm sure you noticed that the author of that article and I [debated back and forth](https://www.matthewathomas.com/programming/2022/06/20/questions-answered.html). He raised the same concern with `ConfigureAwait(false)` as you. And it can indeed lead to mutual exclusion violations when the lock is reentered ([example](https://github.com/matthew-a-thomas/cs-reentrant-async-lock/issues/4#issuecomment-2671711756)). But in my opinion Max could add the following third bullet point under "[The proposed implementation only works if](https://itnext.io/reentrant-recursive-async-lock-is-impossible-in-c-e9593f4aa38a#:~:text=The%20proposed%20implementation%20only%20works%20if)": 3. wrap `ConfigureAwait(false)` code inside an async function (as [recommended here](https://github.com/matthew-a-thomas/cs-reentrant-async-lock#dont-change-the-current-synchronizationcontext-once-youre-in-the-guarded-section)). In my opinion such a bullet list of three items covers enough use cases to say it works in the general case.

Anyway I don't want to distract much more from this issue thread we're on here. For further discussion about the `ReentrantAsyncLock` library I would invite you to open an issue in that repo and tag me there.

**Comment by PolarGoose** [comment]
@matthew-a-thomas,

Ok, amazing. Then, we are on the same page. Sorry I didn't know that you have done such a comprehensive investigation of this problem.

I at least understand why Microsoft hasn't included such primitive to the standard .Net.

**Comment by mqudsi** [comment]
I've published version 3.3.0-preview1 to nuget: https://www.nuget.org/packages/NeoSmart.AsyncLock

There are a couple of bug fixes and a couple of improvements, most important of which is the race condition that was taking place here due to the deferred cleanup (via `Task.Run()`) in the `Dispose()` method. I've also added a hack that abuses deferred awaiting of the result of `WaitAsync()` to hopefully approximate conditional variables to get the ability to "atomically" release one lock while waiting for another (see bddbf6aac3e27628e6b9ac541a08280c7fc7b9fd)

I'd appreciate any testing or confirmation of this build. It's been some time since I last touched this code and it's rather convoluted (because async was never meant to handle recursive locking like this).
